File: C++/strings.cpp
#include <iostream>
#include <string> // required for std::string
using namespace std;

/******* STRINGS ******/

int main() {
    /******* DECLARING STRINGS ******/
    string name; // creates an empty string variable
    cout << "Enter your name: ";
    getline(cin, name); // reads a full line including spaces (cin >> name reads only until first space)

    cout << "Hello " << name << endl;

    /******* STRING FUNCTIONS ******/
    // length() returns number of characters in string
    cout << "Length of name: " << name.length() << endl;

    // substr(start, length) returns substring
    cout << "First 3 letters: " << name.substr(0, 3) << endl;

    // replace(start, length, newString) replaces part of string
    name.replace(0, 3, "Ben"); // replaces first 3 characters with "Ben"
    cout << "Replaced string: " << name << endl;

    /******* COMMON PITFALLS ******/
    // Strings in C++ are objects; they are not arrays of chars like in C.
    // Accessing out-of-bounds indices will cause undefined behavior.
    // Example: name[100] is invalid if name.length() < 101

    return 0;
}

File: Java/Strings.java
package Java;

/******* STRINGS ******/

import java.util.Scanner;

public class Strings {

    public static void main(String[] args){
        Scanner scanner = new Scanner(System.in);

        /******* DECLARING STRINGS ******/
        String name; // empty string
        System.out.print("Enter your name: ");
        name = scanner.nextLine(); // read full line including spaces

        /******* STRING FUNCTIONS ******/
        System.out.println("Hello " + name);
        System.out.println("Length of name: " + name.length()); // number of characters
        System.out.println("Uppercase: " + name.toUpperCase()); // convert to uppercase
        System.out.println("Lowercase: " + name.toLowerCase()); // convert to lowercase

        // substring(start, end) extracts part of string
        if(name.length() >= 3){
            System.out.println("First 3 letters: " + name.substring(0,3));
        }

        // replace(oldChar, newChar) replaces characters
        String replaced = name.replace('a','x'); // replaces all 'a' with 'x'
        System.out.println("After replacement: " + replaced);

        /******* WHY STRINGS ******/
        // Strings are objects in Java, not arrays of characters
        // Use methods like length(), substring(), replace() instead of manual loops
        // Strings are immutable: methods return new strings

        scanner.close();
    }
}
